### PR TITLE
kirkwood: add D-Link DNS-320L support

### PR DESCRIPTION
**Pull request by wigyori**
Dual-slot NAS based on Marvell Kirkwood.

Specifications:
 - Marvell 88F6702 @1GHz
 - 256Mb RAM
 - 128Mb NAND
 - 1x GbE LAN (Marvell 88E1318R)
 - 1x USB 2.0
 - 2x SATA
 - Weltrend WT69P3 ("supervisor" MCU chip)
 - Serial on J2 (115200,8n1)
 - Newer bootROM so kwboot-ing via serial is possible

Notes:
 - The Weltrend MCU is controlled by the package added in utils/dns320l-mcu.
 - The original MAC address is stored in the "mini firmware" image's first 17 bytes.
 - Compared to the original MTD layout, the uImage+rootfs are now stored in a common ubi partition.

Installation:
1. Serial console
 - Connect your levelshifter to the serial console on J2 (refer to the wiki page for pinout)
2. Update u-boot
 - Download the u-boot.kwb image for the device
 - Powercycle the NAS
 - Run "kwboot -b u-boot-dns320l/u-boot.kwb /dev/ttyUSB0 -p"
 - Connect to the serial console with minicom
 - tftp 0x0800000 u-boot-dns320l/u-boot.kwb (Please note that "PHY reset timed out" seems to be customary on kirkwood devices, the egiga0 interface works regardless.)
 - nand erase 0x0 100000
 - nand write 0x0800000 0x0 0x100000
 - reset
3. Install OpenWrt
 - Boot up the initramfs image
 - tftpboot 0x800000 openwrt-kirkwood-generic-dlink_dns320l-initramfs-uImage; bootm 0x800000
 - Download the sysupgrade image and perform sysupgrade

### CODE REVIEW

File: target/linux/kirkwood/image/Makefile [hunk 1]
```diff
@@ -182,6 +182,15 @@ define Device/ctera_c200-v1
 endef
 TARGET_DEVICES += ctera_c200-v1
 
+define Device/dlink_dns320l
+  DEVICE_VENDOR := D-Link
+  DEVICE_MODEL := DNS-320L
+  DEVICE_PACKAGES := kmod-ata-marvell-sata kmod-fs-ext4 kmod-gpio-button-hotplug \
+	kmod-usb-storage kmod-usb-ledtrig-usbport dns320l-mcu
+  SUPPORTED_DEVICES += dns320l
```
Comment (CHKDSK88) [inline target/linux/kirkwood/image/Makefile:190]: Why is this needed?

File: target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+		label = "ubootenv";
+		reg = <0x100000 0x20000>;
+	};
+
+	partition@120000 {
+		label = "ubi";
+		reg = <0x120000 0x6de0000>;
+	};
+
+	partition@6f00000 {
```
Comment (CHKDSK88) [inline target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts:197]: Is any sens to keep it if u-boot is replaced? MAC could be moved to u-boot envs.

File: target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+			pmx_sata0: pmx-sata0 {
+				marvell,pins = "mpp20";
+				marvell,function = "sata1";
+			};
+			pmx_sata1: pmx-sata1 {
+				marvell,pins = "mpp21";
+				marvell,function = "sata0";
+			};
+			pmx_led_red_right_hdd: pmx-led-red-right-hdd {
```
Comment (CHKDSK88) [inline target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts:105]: Recent upstream kirkwood dtses have pinmux squashed.

File: target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+		};
+	};
+
+	gpio-leds {
+		compatible = "gpio-leds";
+		pinctrl-0 = <&pmx_led_usb_blue &pmx_led_red_right_hdd
+			     &pmx_led_red_left_hdd &pmx_led_usb_orange>;
+		pinctrl-names = "default";
+
+		led_usb_blue: blue-usb {
```
Comment (CHKDSK88) [inline target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts:72]: New upstream dtses uses led color and function properties instead label.

File: target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts [hunk 5]
```diff
@@ -0,0 +1,226 @@
+/*
```
Comment (CHKDSK88) [inline target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts:1]: No spdx identifier.

File: target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+ * Device Tree file for DLINK DNS-320L
```
Comment (CHKDSK88) [inline target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts:2]: Are you plan send dts upstream? Debian users form doozan forum will be happy.

**Review by CHKDSK88 — COMMENTED** [review]
Hi @wigyori 
I left some doubts.

File: package/boot/uboot-envtools/files/kirkwood [hunk 7]
```diff
@@ -31,6 +32,9 @@ zyxel,nsa325)
 linksys,ea3500)
 	ubootenv_add_uci_config "/dev/mtd1" "0x0" "0x4000" "0x4000"
 	;;
+dlink,dns-320l)
```
[the remark this answers is not in the document]
Comment (wigyori) [reply package/boot/uboot-envtools/files/kirkwood:35]: Ugh, thanks, gonna fix this.

File: target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts (hunk 2 above)
Comment (wigyori) [reply target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts:197]: I guess so - at least to get the MAC when upgrading from the original firmware. It can be argued that as replacing u-boot will require a serial console, and this assumes that it's at least a power-user who's gonna do it (who can do a setenv ethaddr "bla" in u-boot), but I'm still on the opinion that keeping this partition table would make sense (i.e. if a user wants to move back to the original firmware).

File: target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts (hunk 3 above)
Comment (wigyori) [reply target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts:105]: Will check, thanks.

File: target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts (hunk 6 above)
Comment (wigyori) [reply target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts:2]: Yes, I do - especially as some part of the knowledge came from that forum. :)

File: target/linux/kirkwood/image/Makefile (hunk 1 above)
Comment (wigyori) [reply target/linux/kirkwood/image/Makefile:190]: Which part?

Comment (CHKDSK88) [reply target/linux/kirkwood/image/Makefile:190]: `SUPPORTED_DEVICES` line. In most cases is added because backward compatibility is required.

File: target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts (hunk 6 above)
Comment (wigyori) [reply target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts:3]: @CHKDSK88 I guess all concerns so far have been fixed.

File: target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts (hunk 4 above)
Comment (CHKDSK88) [reply target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts]: I suppose this will be required for upstream. But here I'm not stick to it.

File: target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+			gpios = <&gpio0 25 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "usbport";
+		};
+
+		orange-usb {
+			label = "dns320l:usb:orange";
+			gpios = <&gpio0 26 GPIO_ACTIVE_HIGH>;
+		};
+
+		orange-l_hdd {
```
Comment (CHKDSK88) [inline target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts:76]: From https://docs.kernel.org/devicetree/bindings/dts-coding-style.html :
>Node and property names can use only the following characters:
> - Lowercase characters: [a-z]
> - Digits: [0-9]
>  -  Dash: -

File: target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts (hunk 6 above)
Comment (CHKDSK88) [reply target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts:3]: @wigyori I found one more nit.

**Comment by CHKDSK88** [comment]
LGTM.
`Reviewed-by: Pawel Dembicki <paweldembicki@gmail.com>`

**Comment by wigyori** [comment]
Thanks all for the reviews.

**Comment by xapukserge** [comment]
Hi.
dns320-mcu is not working
cooler had never started 
[kernel.log.txt](https://github.com/user-attachments/files/16026321/kernel.log.txt)
[system.log.txt](https://github.com/user-attachments/files/16026322/system.log.txt)
Although thanks

**Comment by Neustradamus** [comment]
A D-Link DNS-325 @SimplyKot PR is here:
- https://github.com/openwrt/openwrt/pull/17924

**Comment by Neustradamus** [comment]
I see different commits about D-Link DNS-320L:
- kirkwood: add D-Link DNS-320L support (@wigyori committed on Jun 1, 2024): https://github.com/openwrt/openwrt/commit/8619d7af67c28ff974373c9a2937da1025cd81e5
- dns320l-mcu: Mark as nonshared to build in step 1 (@hauke committed on Jul 30, 2024): https://github.com/openwrt/openwrt/commit/9ac50c0aa33f0d0add63a49433147b813b4ce44d
- kirkwood: dns320l: export mcu_power GPIO (@wigyori committed on Oct 28, 2024): https://github.com/openwrt/openwrt/commit/e934e67f9c233b0a458771ed8938d295e33d7de0
- dns320l-mcu: bump to 2024-10-27 (@wigyori committed on Oct 28, 2024): https://github.com/openwrt/openwrt/commit/7ac3f0e96cabb634aaaa779e9f948f5766f288d7

Linked to:
- https://github.com/openwrt/openwrt/pull/16803
- https://github.com/openwrt/openwrt/pull/15433
